sycl/include/sycl/handler.hpp

```diff
@@ -1042,6 +1102,98 @@ class __SYCL_EXPORT handler {
                            item<Dims>, LambdaArgType>>;
   };
 
+  template <int Dims>
+  std::optional<range<Dims>> getRoundedRange(range<Dims> UserRange) {
+    range<Dims> RoundedRange = UserRange;
+    // Disable the rounding-up optimizations under these conditions:
+    // 1. The env var SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING is set.
+    // 2. The kernel is provided via an interoperability method (this uses a
+    // different code path).
+    // 3. The range is already a multiple of the rounding factor.
+    //
+    // Cases 2 and 3 could be supported with extra effort.
+    // As an optimization for the common case it is an
+    // implementation choice to not support those scenarios.
+    // Note that "this_item" is a free function, i.e. not tied to any
+    // specific id or item. When concurrent parallel_fors are executing
+    // on a device it is difficult to tell which parallel_for the call is
+    // being made from. One could replicate portions of the
+    // call-graph to make this_item calls kernel-specific but this is
+    // not considered worthwhile.
+
+    // Perform range rounding if rounding-up is enabled.
+    if (this->DisableRangeRounding())
+      return {};
+
+    // Range should be a multiple of this for reasonable performance.
+    size_t MinFactorX = 16;
+    // Range should be a multiple of this for improved performance.
+    size_t GoodFactor = 32;
+    // Range should be at least this to make rounding worthwhile.
+    size_t MinRangeX = 1024;
+
+    // Check if rounding parameters have been set through environment:
+    // SYCL_PARALLEL_FOR_RANGE_ROUNDING_PARAMS=MinRound:PreferredRound:MinRange
+    this->GetRangeRoundingSettings(MinFactorX, GoodFactor, MinRangeX);
+
+    // In SYCL, each dimension of a global range size is specified by
+    // a size_t, which can be up to 64 bits.  All backends should be
+    // able to accept a kernel launch with a 32-bit global range size
+    // (i.e. do not throw an error).  The OpenCL CPU backend will
+    // accept every 64-bit global range, but the GPU backends will not
+    // generally accept every 64-bit global range.  So, when we get a
+    // non-32-bit global range, we wrap the old kernel in a new kernel
+    // that has each work item peform multiple invocations the old
+    // kernel in a 32-bit global range.
+    auto Dev = detail::getDeviceFromHandler(*this);
+    id<Dims> MaxNWGs = [&] {
+      try {
+        return Dev.get_info<
+            ext::oneapi::experimental::info::device::max_work_groups<Dims>>();
```

Comment:
I might worry about the overhead of repeatedly doing exception-catching on backends that does not support the query.  As an alternative, we could start out by doing `call_nocheck<PiApiKind::piDeviceGetInfo>` ourselves here and use error code directly. It does mean a little more repeat code, but this seems like a potentially critical path.